### ISSUE TITLE
Retrieve partial result sets

### ISSUE THREAD

**Issue by juhlig** [mysql-otp/mysql-otp#94]
The current implementation always retrieves the full result set in a `query` or `execute` call. As I am sometimes dealing with very large result sets, I fear I may run out of memory.

For that reason, it would be nice to have a possibility to fetch only one (or some) resultset rows on demand, let the application process them, then fetch the next, and so on. Somewhat in the style of `match/{3,1}` and `select/{3,1}` in `ets`, you know?

**Comment by zuiderkwast** [comment]
Yes, I'm positive to this feature. An optional `Limit` parameter to all the query and execute functions then, and a `next/1` to get the next chunk? An the return values of query and execute have to be extended with a continuation. Or perhaps we can only have one such chunked result waiting at a time. Then we don't need a continuation. Just a `next/0`.

How to combine chunking with calls that may returns multiple different result sets? Perhaps we don't need to support that...

```Erlang
{ok, [ResultFoo, ResultBar]} = mysql:query(Pid, "SELECT * foo; SELECT * from bar;").
```

Feel free to implement it!

**Comment by juhlig** [comment]
I think there are a few issues with a chunked retrieval that complicate integration into the current functions.

Namely, the resultset (including the coldefs) of a query is only available to the client when the resultset of a previous query has been fetched completely.

As an illustration, if you issue a query like `SELECT * FROM foo` and don't fetch anything yet, this is what is in the receive buffer:
`[colcount] [coldef] ... [eof] [row] ... [eof]`
If you then issue another query (or use a multi statement like in your comment) like `SELECT * FROM bar`, this is what is in the receive buffer:
`[colcount] [coldef] ... [eof] [row] ... [eof] [colcount] [coldef] ... [eof] [row] ... [eof]`

By using chunked fetching, a client must operate in a more sequential fashion and obey to some additional restrictions, and/or use multiple connections. I don't see how the library could sensibly handle this, so it is in the responsibility of the client.

Specifically, it is not possible (with 1 connection) to read a resultset chunked and update something, even if totally unrelated, or fetch additional information while processing it. The C API handles this dilemma in a straightforward manner: it just forbids it; if you want it, you have to use a separate connection.

Also, multi statements are not really possible, at least not in a manner that would return something like the current `query` function does. The only way I see to achieve this would be by using two continuations, one for "more rows in a resultset", one for "more resultsets in response".

Third, it is not possible to discard a partially read resultset as such (like, you fetched a few rows and are not interested in the rest), it must be fetched fully, though this can be hidden behind a function that simply does just that.

**Comment by zuiderkwast** [comment]
To forbid other queries until the previous result set has been fully fetched, just like the C API does, makes good sense to me. We can also forbid multiple queries in the same string when chunked retrieval is used.

**Comment by juhlig** [comment]
The first part, disallowing another `query` call when a result set is still active, seems not to be too complicated. But how would you address the second, disallowing multi statements? I mean, without parsing the query string? As far as I know, it is not possible to change capabilities (CLIENT_MULTI_STATEMENTS) outside of the connection/handshake phase.

Another question before I start digging deeper into this is, what OTP versions do you want to support? From my browsing in the code, it seems that it should as yet work with versions at least as far back as OTP-16. Is that a requirement, then?

**Comment by juhlig** [comment]
Ah, it seems I was mistaken. It _should_ be possible to enable and disable multi statements with `COM_SET_OPTION`.

**Comment by zuiderkwast** [comment]
For multi statements in the same query call, it can be just disallowed by documentation. The programmer should know what queries are sent. If there is a second result set coming from the server after we have done a chunked retreival of the first result set, we can just crash. I think that is fine, as long as it is stated in documentation that it is not supported.

Yes, currently OTP 16 and MySQL 4.1 are supported. I think it is good that it can work in really old projects as well.

**Comment by juhlig** [comment]
Another option would be to decide on connect whether results shall be fetched chunked or all-at-once, thereby enabling or disabling the multi statements capability flag implicitly, though that may introduce some problems I'm not aware of. If a purely documentary restriction is fine with you, it's fine with me :)

So I'll lay out a plan over the weekend and start playing some time next week then.

**Comment by juhlig** [comment]
Sorry, it looks like I won't get around to work on this anytime soon after all :(

**Comment by zuiderkwast** [comment]
OK, no problem! Currently, the way to chunk the data is to use `SELECT ... LIMIT m OFFSET n;`.

**Comment by juhlig** [comment]
Yes, it is a workaround, but there are issues with that approach which could be addressed with chunked retrieval, namely:
* The query needs to be issued multiple times, and the results will include the columns header which needs to be parsed each time. The performance impact here should be light, but still it is there, and will add up if the overall result is large.
* The query needs to be executed multiple times on the server, which will be a serious burden with complex, time-consuming queries. The performance impact here may be considerable, and adds up just the same as in the previous point. (I don't know how smart MySQLs cache is in that regard, though)
* It only works reliably if used in a transaction, with isolation level higher than `READ COMMITTED`. Otherwise, some rows would be missed if, between the chunking queries, rows that were already read in this fashion are being deleted.

Aside from the memory issue which prompted me to start this ticket, those are other points that make chunked retrieval desirable.

**Comment by juhlig** [comment]
I have been thinking about this again lately. What if, instead of retrieving results partially by follow-up calls to retrieve the next row, we created the possibility to specify a callback function that would then be called for each row, after retrieval? Said function could then do some processing, and either throw it away or return a transformation of it, to be collected and returned to the caller at the end? Some kind of `filtermap` over the result, if you want. What do you think?

**Comment by zuiderkwast** [comment]
Yes! One problem with that is that the mysql connection runs in its own process. How about you specify process ID which gets the results by message passing?

**Comment by juhlig** [comment]
A function would be much more flexible, I think. I don't see its being executed in the mysql connection process as a problem, quite the opposite, actually. I'll try to elaborate and highlight the problems I see with a send-it-off approach and the benefits of a simple function :)

<hr />

For sending off the result rows to another process, you have to invent some kind of "protocol", ie in what form the messages should be sent. A tagged tuple is the obvious choice, ok. But you also have to decide (by implementation in mysql-otp) if you want it to be sent in a synchronous (send a row, wait for an ack or request for more; in which case you would have to invent conventions for the rreplies, also) or asynchronous (just send off the rows as fast as they come) fashion, or create another option by which the user can decide which he wants. However, in both cases he is restrained by the protocol.

The asynchronous approach has the benefit that, even if the receiver process dies, it does not concern the connection process, it will just keep sending messages to a non-existing process. The downside is that it may easily clog the message queue of said process with many row messages, even if it is processing them quite fast. Aside from that, I believe message ordering is not guaranteed if the sending and receiving process are on different nodes, which may or may not be a problem for a receiver making that assumption.

The synchronous approach has the benefits and drawbacks of the asynchronous approach reversed, more or less. The requirement to ack or request more acts as a throttle on how fast rows are sent to it, and also guarantees message ordering, but if it dies or messages get lost, there will be no acks or requests for more and the connection process gets stuck, unless you go to the length of implementing a `gen`-style monitoring around the messaging.

The third option, let the user specify if he wants synchronous or asynchronous, is a hassle, as you basically implement both approaches and let the user have basically the choice between two fixed sets of benefit/drawback tradeoffs.

<hr />

The benefit of a simple function that gets called for each result row is that it can do all of the above and more, at the users choice. The function may be implemented to send the row it gets to another process in whatever format the user wants the messages to be, and the user can pick if he wants to do it simply with the `!` operator, with `gen_*:call` or `gen_*:cast` or whatever, thereby if it is synchronous or asynchronous is wholly up to the function implementation; it could also just write the rows to disk; or do all of them and other things.

The function implementation also implicitly determines the throttling effect on the connection process, since only when it returns the next row will be fetched. If it returns quickly (like, with just asynchronously sending off the row somewhere), the connection process will go through the result quickly. If it does complicated stuff instead, the retrieval of the result will be slower, too, but that's just as well since in that case the result couldn't be processed right away, anyway.

The important aspect, I think, is that it all is at the freedom (and responsibility) of the user. mysql-otp doesn't have to implement, provide for or guard against much. Just call the function, let it do whatever it does, it's of no concern.

Downside to it all is only if the function does something funny and crashes the connection process, but that would then be in the responsibility of the user, also.

**Comment by zuiderkwast** [comment]
Good points!

For large restults, it's actually good that it runs in the mysql-otp process for performance reasons: no extra copying.

So, I assume the function should take one row at a time, right?

If it's an unary function, it gets the row. If it's 2-ary function, it gets the column names and the row. How about that?

For the API, something like this?

```Erlang
-type query_callback() :: fun((list(), list()) -> any()) | fun((list()) -> any()).
-spec mysql:async_query(Pid, Query [, Params] [, Timeout], query_callback()) -> ok | {error, server_reason()}.
-spec mysql:async_execute(Pid, Stmt, [, Params] [, Timeout], query_callback()) -> ok | {error, server_reason()}.
```

Btw, should it be asynchronous? I.e. should the function be called `async_query` and return `ok` directly and the callback be called later? In that case, we need some way to tell when all the rows have been fetched.

If we instead make it synchronous, it would hang and return `ok` when all the rows have been fetched an the callback has returned. It can perhaps be called just `query` and `execute` and we treat the callback as an optional parameter. Any preferences?

**Comment by juhlig** [comment]
This is how it works now:
1. Fetch entire resultset
2. Parse resultset
3. Get column definitions
4. Decode all rows
5. If there are more resultsets, repeat from 1.

What I am imagining is this:
* For each resultset,
  1. Get column definitions
  2. Fetch 1 row
  3. Parse/decode row
  4. Pass row to function. There are two options here:
    * [`filtermap`](http://erlang.org/doc/man/lists.html#filtermap-2) style. In this case, the column definitions should be passed to the function, also, since this is the only way for that function to know which resultset the row belongs. The function should return either:
      * `false`, meaning to not include the row in the final resultset returned to the caller of `query`
      * `true`, meaning to include the row unchanged
      * `{true, Value}`, meaning to include `Value` instead of the original row
    * [`foldl`](http://erlang.org/doc/man/lists.html#foldl-3) style. In this case, I would suggest to send tagged tuples to the function, like `{coldefs, ColumnDefinitions}`, `{row, Row}`, and `eof` (end of resultset), along with an accumulator. The function can decide to keep the coldefs in the accumulator if it needs them, and strip them out again when it receives the `eof` or do whatever post-processing it wants. What gets sent back to the caller of `query` would be whatever is resturned from the function call with `eof` in it.

The `foldl` approach is more powerful than the `filtermap` one, as the former allows the entire resultset to be transformed into something else, while the latter can transform (or filter out) only individual rows from the resultset. On the other hand, I'm not sure a `foldl` approach would not be over the top, it's more complicated for the user.

Anyway, the API for a `filtermap` solution would be:
```erlang
-type query_callback() :: fun(list(), list()) -> boolean() | {true, term()}.
```
For a `foldl`approach, it would be
```erlang
-type query_callback() :: fun({coldefs | row, list()} | eof, term()) -> term().
```

<hr />

Finally, I would say it definitely should _not_ be asynchronous. I don't see any benefit in that, and it would make things much more complicated ("... some way to tell when all the rows have been fetched."). Specifically, the connection is unusable until it has finished, anyway. If the calling process wants to do something else in the meantime, it can (and should) run `query` in a process it spawned itself, as it would do right now.

**Comment by juhlig** [comment]
For illustration purposes, I uploaded an (extremely hacky and ugly, so shield your eyes) implementation of this: https://github.com/juhlig/mysql-otp/tree/query_filtermap_hacky.

The only thing that actually works are an un-parameterized calls to `mysql:query/{2,3,4}` (the rest, I probably broke), like
```erlang
mysql:query(Conn, Query, FilterMapFun)
```

Here are some examples, with a query that selects three columns `a`, `b` and `c` from a table:

This one keeps only rows where `a` is <10, and discards the otehrs:
```erlang
mysql:query(
    Conn,
    <<"SELECT a, b, c FROM foo">>,
    fun (_, [A|_]) -> A<10 end
)
```

This one transforms the list that is a row to a tuple:
```erlang
mysql:query(
    Conn,
    <<"SELECT a, b, c FROM foo">>,
    fun (_, Row)) -> {true, list_to_tuple(Row)} end
)
```

This one keeps all rows where `a` is <10 as-is, transforms all rows where `a` is >=10 and <20 to tuples, and discards the rest:
```erlang
mysql:query(
    Conn,
    <<"SELECT a, b, c FROM foo">>,
    fun
        (_, [A|_]) when A<10 -> true;
        (_, Row=[A|_]) when A<20 -> {true, list_to_tuple(Row)};
        (_, _) -> false
    end
)
```

One thing I noticed is that, while the column definitions are returned as they are returned from the query, they might not match the rows any more after they have been mapped (the function might add/drop elements to/from the list, change the ordering, exchange them for something different, etc). Not sure if that's a problem in the real world (the user implementing the function should know about this), but it isn't easily mitigated.

**Comment by zuiderkwast** [comment]
I like these ideas and I don't really see any problem with any of this, so feel free to continue this!

I like the filtermap interface better than fold. The filtermap interface can be used as a simple map, to convert the data to another format (records, maps, etc) or only for filtering (which should be done using SQL code when possible, but maybe not always possible) and side effects such as IO, then just returning false for each row. The interface is simple. The fold interface is too complicated I think.

It would also be possible to support both 1-ary fun (row only) and 2-ary fun (column names and row). In most cases, there is only one result set per query, so in those cases the 1-ary fun variant would be more user-friendly.

**Comment by juhlig** [comment]
Resolved with #104
